File: samlang-checker/src/commonMain/kotlin/samlang/checker/StatementTypeChecker.kt
package samlang.checker

import samlang.ast.common.Type
import samlang.ast.common.TypeDefinitionType
import samlang.ast.lang.Expression
import samlang.ast.lang.Pattern
import samlang.ast.lang.Statement
import samlang.ast.lang.Statement.Val
import samlang.ast.lang.StatementBlock
import samlang.errors.TupleSizeMismatchError
import samlang.errors.UnexpectedTypeKindError
import samlang.errors.UnresolvedNameError

internal class StatementTypeChecker(
    private val accessibleGlobalTypingContext: AccessibleGlobalTypingContext,
    private val errorCollector: ErrorCollector,
    private val expressionTypeChecker: ExpressionTypeCheckerWithContext
) {
    fun typeCheck(
        statementBlock: StatementBlock,
        expectedType: Type,
        localContext: LocalTypingContext
    ): StatementBlock = localContext.withNestedScope {
        typeCheckInNestedScope(
            statementBlock = statementBlock,
            expectedType = expectedType,
            localContext = localContext
        )
    }

    private fun typeCheckInNestedScope(
        statementBlock: StatementBlock,
        expectedType: Type,
        localContext: LocalTypingContext
    ): StatementBlock {
        val checkedStatements = mutableListOf<Statement>()
        for (statement in statementBlock.statements) {
            checkedStatements += typeCheck(
                statement = statement,
                localContext = localContext
            )
        }
        val expression = statementBlock.expression
        val checkedExpression = if (expression != null) {
            expressionTypeChecker.typeCheck(
                expression = expression,
                expectedType = expectedType
            )
        } else {
            // Force the type checker to resolve expected type to unit.
            expressionTypeChecker.typeCheck(
                expression = Expression.Panic(
                    range = statementBlock.range,
                    type = Type.unit,
                    expression = Expression.Literal.ofString(range = statementBlock.range, value = "")
                ),
                expectedType = expectedType
            )
            null
        }
        return statementBlock.copy(statements = checkedStatements, expression = checkedExpression)
    }

    private fun typeCheck(statement: Statement, localContext: LocalTypingContext): Statement =
        when (statement) {
            is Val -> typeCheckVal(statement = statement, localContext = localContext)
        }

    private fun typeCheckVal(statement: Val, localContext: LocalTypingContext): Val {
        val (_, pattern, typeAnnotation, assignedExpression) = statement
        val checkedAssignedExpression = expressionTypeChecker.typeCheck(
            expression = assignedExpression,
            expectedType = typeAnnotation
        )
        val betterStatement = statement.copy(assignedExpression = checkedAssignedExpression)
        val checkedAssignedExprType = checkedAssignedExpression.type
        val checkedPattern = when (pattern) {
            is Pattern.TuplePattern -> {
                val tupleType = checkedAssignedExprType as? Type.TupleType ?: kotlin.run {
                    errorCollector.add(
                        compileTimeError = UnexpectedTypeKindError(
                            expectedTypeKind = "tuple",
                            actualType = checkedAssignedExprType,
                            range = assignedExpression.range
                        )
                    )
                    return betterStatement
                }
                val expectedSize = tupleType.mappings.size
                val actualSize = pattern.destructedNames.size
                if (expectedSize != actualSize) {
                    errorCollector.add(
                        compileTimeError = TupleSizeMismatchError(
                            expectedSize = expectedSize,
                            actualSize = actualSize,
                            range = assignedExpression.range
                        )
                    )
                }
                pattern.destructedNames.zip(tupleType.mappings).asSequence().mapNotNull { (nameWithRange, t) ->
                    val (name, nameRange) = nameWithRange
                    if (name == null) null else Triple(first = name, second = nameRange, third = t)
                }.forEach { (name, nameRange, elementType) ->
                    localContext.addLocalValueType(name = name, type = elementType) {
                        errorCollector.reportCollisionError(name = name, range = nameRange)
                    }
                }
                pattern
            }
            is Pattern.ObjectPattern -> {
                val identifierType = checkedAssignedExprType as? Type.IdentifierType ?: kotlin.run {
                    errorCollector.add(
                        compileTimeError = UnexpectedTypeKindError(
                            expectedTypeKind = "identifier",
                            actualType = checkedAssignedExprType,
                            range = assignedExpression.range
                        )
                    )
                    return betterStatement
                }
                val fieldMappingsOrError = ClassTypeDefinitionResolver.getTypeDefinition(
                    identifierType = identifierType,
                    context = accessibleGlobalTypingContext,
                    typeDefinitionType = TypeDefinitionType.OBJECT,
                    errorRange = assignedExpression.range
                )
                val (fieldNames, fieldMappings) = when (fieldMappingsOrError) {
                    is Either.Left -> fieldMappingsOrError.v
                    is Either.Right -> {
                        errorCollector.add(compileTimeError = fieldMappingsOrError.v)
                        return betterStatement
                    }
                }
                val fieldOrderMapping = fieldNames.asSequence().mapIndexed { index, name -> name to index }.toMap()
                val orderedDestructuredName = pattern.destructedNames.map { destructuredName ->
                    val (originalName, _, renamedName, fieldRange) = destructuredName
                    val (fieldType, isPublic) = fieldMappings[originalName] ?: kotlin.run {
                        errorCollector.add(UnresolvedNameError(unresolvedName = originalName, range = fieldRange))
                        return betterStatement
                    }
                    if (identifierType.identifier != accessibleGlobalTypingContext.currentClass && !isPublic) {
                        errorCollector.add(UnresolvedNameError(unresolvedName = originalName, range = fieldRange))
                        return betterStatement
                    }
                    val nameToBeUsed = renamedName ?: originalName
                    localContext.addLocalValueType(name = nameToBeUsed, type = fieldType) {
                        errorCollector.reportCollisionError(name = nameToBeUsed, range = fieldRange)
                    }
                    val order = fieldOrderMapping[originalName] ?: error(message = "Bad field!")
                    destructuredName.copy(fieldOrder = order)
                }.sortedBy { it.fieldOrder }
                pattern.copy(destructedNames = orderedDestructuredName)
            }
            is Pattern.VariablePattern -> {
                val (p, n) = pattern
                localContext.addLocalValueType(name = n, type = checkedAssignedExprType) {
                    errorCollector.reportCollisionError(name = n, range = p)
                }
                pattern
            }
            is Pattern.WildCardPattern -> pattern
        }
        return betterStatement.copy(pattern = checkedPattern)
    }
}

File: samlang-core-checker/statement-type-checker.ts
import { Type, unitType } from 'samlang-core-ast/common-nodes';
import {
  SamlangExpression,
  Pattern,
  ObjectPatternDestucturedName,
  SamlangValStatement,
  StatementBlock,
  SourceExpressionVariable,
} from 'samlang-core-ast/samlang-nodes';
import type { FieldType } from 'samlang-core-ast/samlang-toplevel';
import type { ModuleErrorCollector } from 'samlang-core-errors';
import { checkNotNull, zip, LocalStackedContext, assert } from 'samlang-core-utils';

import type { AccessibleGlobalTypingContext } from './typing-context';

export default class StatementTypeChecker {
  constructor(
    private readonly accessibleGlobalTypingContext: AccessibleGlobalTypingContext,
    private readonly errorCollector: ModuleErrorCollector,
    private readonly typeCheckExpression: (
      expression: SamlangExpression,
      expectedType: Type
    ) => SamlangExpression
  ) {}

  readonly typeCheck = (
    { range, statements, expression }: StatementBlock,
    expectedType: Type,
    localContext: LocalStackedContext<Type>
  ): StatementBlock =>
    localContext.withNestedScope(() => {
      const checkedStatements = statements.map((statement) =>
        this.typeCheckValStatement(statement, localContext)
      );
      if (expression != null) {
        const checkedExpression = this.typeCheckExpression(expression, expectedType);
        return { range, statements: checkedStatements, expression: checkedExpression };
      }
      // Force the type checker to resolve expected type to unit.
      this.typeCheckExpression(
        SourceExpressionVariable({
          range,
          type: unitType,
          associatedComments: [],
          name: '_',
        }),
        expectedType
      );
      return { range, statements: checkedStatements };
    });

  private typeCheckValStatement(
    statement: SamlangValStatement,
    localContext: LocalStackedContext<Type>
  ): SamlangValStatement {
    const { range, pattern, typeAnnotation, assignedExpression } = statement;
    const checkedAssignedExpression = this.typeCheckExpression(assignedExpression, typeAnnotation);
    const checkedAssignedExpressionType = checkedAssignedExpression.type;
    let checkedPattern: Pattern;
    switch (pattern.type) {
      case 'TuplePattern': {
        if (checkedAssignedExpressionType.type !== 'TupleType') {
          this.errorCollector.reportUnexpectedTypeKindError(
            assignedExpression.range,
            'tuple',
            checkedAssignedExpressionType
          );
          return {
            ...statement,
            typeAnnotation: assignedExpression.type,
            assignedExpression: checkedAssignedExpression,
          };
        }
        const expectedSize = checkedAssignedExpressionType.mappings.length;
        const actualSize = pattern.destructedNames.length;
        if (expectedSize !== actualSize) {
          this.errorCollector.reportTupleSizeMismatchError(
            assignedExpression.range,
            expectedSize,
            actualSize
          );
        }
        const checkedDestructedNames = zip(
          pattern.destructedNames,
          checkedAssignedExpressionType.mappings
        ).map(([{ name, range: nameRange }, elementType]) => {
          if (name != null) {
            localContext.addLocalValueType(name, elementType, () =>
              this.errorCollector.reportCollisionError(nameRange, name)
            );
          }
          return { name, type: elementType, range: nameRange };
        });
        checkedPattern = { ...pattern, destructedNames: checkedDestructedNames };
        break;
      }

      case 'ObjectPattern': {
        if (checkedAssignedExpressionType.type !== 'IdentifierType') {
          this.errorCollector.reportUnexpectedTypeKindError(
            assignedExpression.range,
            'identifier',
            checkedAssignedExpressionType
          );
          return {
            ...statement,
            typeAnnotation: assignedExpression.type,
            assignedExpression: checkedAssignedExpression,
          };
        }
        const fieldMappingsOrError = this.accessibleGlobalTypingContext.resolveTypeDefinition(
          checkedAssignedExpressionType,
          'object'
        );
        let fieldNamesMappings: {
          readonly fieldNames: readonly string[];
          readonly fieldMappings: Readonly<Record<string, FieldType>>;
        };
        assert(
          fieldMappingsOrError.type !== 'IllegalOtherClassMatch',
          'We match on objects here, so this case is impossible.'
        );
        switch (fieldMappingsOrError.type) {
          case 'Resolved':
            fieldNamesMappings = {
              fieldNames: fieldMappingsOrError.names,
              fieldMappings: fieldMappingsOrError.mappings,
            };
            break;
          case 'UnsupportedClassTypeDefinition':
            this.errorCollector.reportUnsupportedClassTypeDefinitionError(
              assignedExpression.range,
              'object'
            );
            return {
              ...statement,
              typeAnnotation: assignedExpression.type,
              assignedExpression: checkedAssignedExpression,
            };
        }
        const { fieldNames, fieldMappings } = fieldNamesMappings;
        const fieldOrderMapping = Object.fromEntries(
          fieldNames.map((name, index) => [name, index])
        );
        const destructedNames: ObjectPatternDestucturedName[] = [];
        for (let i = 0; i < pattern.destructedNames.length; i += 1) {
          const destructedName: ObjectPatternDestucturedName = checkNotNull(
            pattern.destructedNames[i]
          );
          const { fieldName: originalName, alias: renamedName, range: fieldRange } = destructedName;
          const fieldInformation = fieldMappings[originalName];
          if (fieldInformation == null) {
            this.errorCollector.reportUnresolvedNameError(fieldRange, originalName);
            return {
              ...statement,
              typeAnnotation: assignedExpression.type,
              assignedExpression: checkedAssignedExpression,
            };
          }
          const { isPublic, type: fieldType } = fieldInformation;
          if (
            checkedAssignedExpressionType.identifier !==
              this.accessibleGlobalTypingContext.currentClass &&
            !isPublic
          ) {
            this.errorCollector.reportUnresolvedNameError(fieldRange, originalName);
            return {
              ...statement,
              typeAnnotation: assignedExpression.type,
              assignedExpression: checkedAssignedExpression,
            };
          }
          const nameToBeUsed = renamedName?.[0] ?? originalName;
          localContext.addLocalValueType(nameToBeUsed, fieldType, () =>
            this.errorCollector.reportCollisionError(fieldRange, nameToBeUsed)
          );
          const fieldOrder = checkNotNull(fieldOrderMapping[originalName]);
          destructedNames.push({ ...destructedName, type: fieldType, fieldOrder });
        }
        checkedPattern = { range, type: 'ObjectPattern', destructedNames };
        break;
      }

      case 'VariablePattern':
        localContext.addLocalValueType(pattern.name, checkedAssignedExpressionType, () =>
          this.errorCollector.reportCollisionError(pattern.range, pattern.name)
        );
        checkedPattern = pattern;
        break;

      case 'WildCardPattern':
        checkedPattern = pattern;
        break;
    }
    return {
      ...statement,
      typeAnnotation: assignedExpression.type,
      assignedExpression: checkedAssignedExpression,
      pattern: checkedPattern,
    };
  }
}
